File: codes/swift/p970.swift
import Foundation
import ComplexModule

func lambertW(_ z: Complex<Double>, k: Int) -> Complex<Double> {
    let r = z.length
    let theta = z.phase
    var w = Complex(log(r), theta + 2*Double.pi*Double(k))

    for _ in 0..<50 {
        let ew = w.exp
        let f  = w*ew - z
        let fp = ew*(w + Complex(1,0))
        w -= f/fp
    }
    return w
}

File: codes/rust/p970.rs
use num_complex::Complex;
use std::f64::consts::PI;

fn lambertw(z: Complex<f64>, k: i32) -> Complex<f64> {
    let r = z.norm();
    let theta = z.arg();
    let mut w = Complex::new(r.ln(), theta + 2.0*PI*(k as f64));

    for _ in 0..50 {
        let ew = w.exp();
        let f  = w*ew - z;
        let fp = ew*(w + Complex::new(1.0,0.0));
        w -= f/fp;
    }
    w
}
